atlasdb-commons/src/main/java/com/palantir/common/compression/GzipCompressingInputStream.java

```diff
@@ -0,0 +1,117 @@
+/*
+ * (c) Copyright 2018 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.palantir.common.compression;
+
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.SequenceInputStream;
+import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
+import java.util.Arrays;
+import java.util.Enumeration;
+import java.util.Iterator;
+import java.util.function.Supplier;
+import java.util.zip.CRC32;
+import java.util.zip.CheckedInputStream;
+import java.util.zip.Deflater;
+import java.util.zip.DeflaterInputStream;
+
+import com.google.common.io.CountingInputStream;
+
+
+public class GzipCompressingInputStream extends SequenceInputStream {
+    public GzipCompressingInputStream(InputStream in) throws IOException {
+        this(in, 512);
+    }
+
+    public GzipCompressingInputStream(InputStream in, int bufferSize) throws IOException {
+        super(new GzipStreamEnumeration(in, bufferSize));
+    }
+
+    protected static class GzipStreamEnumeration implements Enumeration<InputStream> {
```

Comment:
Implementing `Enumeration` for 3 elements is the Java 1.0 equivalent of implementing an `Iterable` for the same thing.

What you do here is create a list, turn it into an iterator, and then implement an enumeration around the iterator. Instead, just do `Collections.enumeration(theListYouHad)` in order to convert the list into an enumeration.

Reply:
You still don't need to implement enumeration for that, though. That said, using suppliers makes sense now.

Reply:
So explicitly, I'm proposing you write something like:

```java
List<Supplier<InputStream>> streams = ImmutableList.of(this::createHeaderStream, () -> in, () -> countingStream, () -> createTrailerStream(crc, counting));
Enumeration<InputStream> inputStream = Collections.enumeration(Collections2.transform(streams, Supplier::get)); 
```

Reply:
But as the GzipCompressingInputStream is extending from the SequenceInputStream, the first call of the constructor needs to be a super(). So I cannot initialize the inputstream enum above in the  GzipCompressingInputStream constructor.